kolibri/core/device/soud.py

```diff
@@ -0,0 +1,425 @@
+        and sync_queue.attempt_at > attempt_execute_window()
+    ):
+        # if we have a queue id and it's not time to recheck, don't re-request. this protects
+        # against potential flippant network connections
+        return False
+    # otherwise, we're good to go
+    return True
+
+
+def handle_sync_request_response(context, sync_queue, response):
+    data = response.json()
+    if data["status"] == SyncQueueStatus.Ready:
+        logger.info("{} SoUD sync ready".format(context))
+        sync_queue.status = SyncQueueStatus.Ready
+        sync_queue.reset_next_attempt(0)  # now
+        sync_session = find_client_sync_session(context)
+        if sync_session:
+            sync_queue.sync_session_id = sync_session.id
+    elif data["status"] == SyncQueueStatus.Queued:
+        logger.info("{} SoUD sync queued".format(context))
+        sync_queue.status = SyncQueueStatus.Queued
+        sync_queue.reset_next_attempt(int(data.get("keep_alive", 0)))
+    else:
+        logger.warning(
+            "{} Unknown response action for SoUD sync request | {}".format(
+                context, data
+            )
+        )
+        sync_queue.status = SyncQueueStatus.Pending
+
+
+def request_sync(context, network_location=None):
+    """
+    Request a sync from the device with the given instance_id.
+    """
+    from kolibri.core.auth.tasks import queue_soud_sync_cleanup
+
+    sync_queue = context.sync_queue
+    if not validate_sync_queue_for_sync_request(sync_queue):
+        return
+
+    network_location = network_location or context.network_location
+
+    # the SoUD syncing is currently dependent on network discovery, but as long as we have a
+    # network location, we can request a sync
+    if network_location is None:
+        logger.info("{} Network location unavailable".format(context))
+        if sync_queue.sync_session_id:
+            # remove sync session
+            queue_soud_sync_cleanup(sync_queue.sync_session_id)
+            sync_queue.sync_session_id = None
+        sync_queue.increment_and_backoff_next_attempt()
+        sync_queue.save()
+        return
+
+    try:
+        with SoudNetworkClient.build_from_network_location(network_location) as client:
+            if sync_queue.status == SyncQueueStatus.Pending:
+                # when Pending, we are not sure if we have a sync queue record on the remote server,
+                # so we request to create one
+                response = client.request_sync_queue(context)
+            else:
+                # when not Pending, we request to update its status, which the server should return
+                response = client.request_sync_queue_status(context)
+    except (NetworkClientError, NetworkLocationNotFound):
+        logger.warning(
+            "{} Unable to request SoUD sync from unavailable network location".format(
+                context
+            )
+        )
+        sync_queue.increment_and_backoff_next_attempt()
```

Comment:
Should we be backing off in this case, or waiting for the device to come back online again and then requeueing instead?

Reply:
I want to say that we could skip the backoff, but I think there are scenarios where we do get some benefit here in the time being, particularly with static locations that may not be on the same subnet. During the next run, if the network location doesn't exist or is marked with a connection error, it won't make any requests, but continue to backoff until it exceeds the max attempts